### DISCUSSION TITLE
Wayland Native Backend?

### DISCUSSION THREAD

**Discussion by leroycep** [ghostty-org/ghostty#3574] [Feature Requests, Ideas] [answered]
You might think I'm something of a zealot, but I don't consider GTK or X11 to be the native toolkit of Linux. Several of the decisions the GTK project has made leave a sour taste in mouth, including some of the ones around wayland-protocols. X11 is okay but I don't expect it will receive much development work going forward. I consider going through something like GTK or GLFW as frameworks that introduce unnecessary cross-platform abstractions, at least for an app that is trying to be native.

(Side note: I also don't consider Linux to have a "system libc". `glibc` is a very popular cross platform libc installed by default on many, but not all, Linux distros. `musl` is another option, or simply having no libc at all. Compare this to Windows and MacOS (or the BSDs), where there is a single system library that, 1. will be installed, and 2. must be used to access the system. This paragraph is mainly added to add some additional context when one is considering how extreme my opinions are.)

Anyway, I don't know how far along `libghostty` is, but would a Wayland native backend be considered? Or would something like this be left to a third party project to implement?

(Also, I realize asking for native support for a subset of Linux before Windows support has been implemented a bit silly 😅)

**Comment by mitchellh** [comment]
We support Wayland without any X11 compatibility perfectly fine via GTK. Wayland being the protocol it is doesn't provide the features we need in our toolkit: widgets (and I wouldn't expect Wayland too, I understand its place in the layer cake). We're comparing Apples and Oranges here.

libghostty could definitely be used to build a non-GTK build that only runs on Wayland but it wouldn't be an official Ghostty app because "native UI" is part of our goals. But that's exactly what libghostty is for -- to enable the community to do stuff like this 😄

**Comment by jokeyrhyme** [comment]
It seemed like this was two separate questions (wayland? answered; GTK alternatives? discussed), so I'm going to resume the second part (apologies if this should have been a separate GitHub Discussion)...

Taking a look at https://distrowatch.com/dwres.php?resource=major and identifying the default environments (when present):

| distribution | desktops (toolkit) |
|---|---|
| Linux Mint | Cinnamon (GTK), MATE (GTK), Xfce (GTK) |
| Ubuntu | GNOME (GTK) |
| MX Linux | Fluxbox (?), KDE (Qt), Xfce (GTK) |
| Slackware | KDE (Qt) |
| Debian | GNOME (GTK) |
| Fedora | GNOME (GTK) |
| OpenSUSE | GNOME (GTK), KDE (Qt), Xfce (GTK) |
| RedHat | GNOME (GTK) |

And I'm sure this was investigated early on as part of deciding that GTK was the "native" toolkit on Linux, as irksome as this may be to me personally and others, haha

Just for representation, I'd like to mention that Linux hardware vendors ship their devices with Raspberry Pi OS (using PIXEL desktop, not sure which toolkit this is) and Pop_OS! (COSMIC, COSMIC toolkit)

Personally, I'm a huge fan of COSMIC, so I'm very curious to see how COSMIC Terminal and Ghostty mature and how well the "native" integration in Ghostty end up working on that desktop environment

**Reply by mitchellh** [reply, marked as answer]
> And I'm sure this was investigated early on as part of deciding that GTK was the "native" toolkit on Linux, as irksome as this may be to me personally and others, haha

That's right. It was a pragmatic choice.

Alternate apprts for Linux aren't currently planned officially. This is a perfect use case for libghostty when available, or a fork in the mean time.

**Reply by jokeyrhyme** [reply]
I did think of documentation-solution for these expectations

What if you listed "Linux (GNOME)" separate from "Linux", in the support/feature matrix, so you could explicitly say that "native"-ness is a goal/feature on "Linux (GNOME)" but is not so for the rest of "Linux"?

**Reply by leroycep** [reply]
I would be for this change, but mostly in the context of eventually adding other Linux backends. If no other official Linux backends are planned, I don't think this adds anything that isn't already said by "Linux support via GTK".

**Reply by jokeyrhyme** [reply]
It'd at least appease the pedants among us regarding the "GTK is not 'native'" sentiment

Us engineering and Linux folks do love to bike-shed RE: accurate use of language :P

**Comment by 00-kat** [comment]
A member of the community has made one now: https://github.com/gabydd/wraith.
